### DISCUSSION TITLE
Bridge local MQTT broker to SAJ server

### DISCUSSION THREAD

**Discussion by paolosabatino** [paolosabatino/saj-mqtt-ha#4] [General]
This post comes from [this issue](https://github.com/paolosabatino/saj-mqtt-ha/issues/2), in particular it has been posted by @h3llrais3r

> My inverter is an H1-3.6K-S2 with firmware 1.361

Ok, I thought the infamous firmware was for the inverter... but it's for the battery I assume?
My inverter is the H1-5K-S2 with firmware 1.202.
For the battery I can see a software version, which seems to be 1.152

> Can you explain how to do that bridging?

Based on https://solar-assistant.io/help/home-assistant/broker-setup

- Add DNS rewrite in adguard for `mqtt.saj-solar.com` to your mqtt ip address (I'm running the mosquitto addon in HA, so for me it's the HA ip)
- Add mqtt bridge configuration in HA mqtt addon configuration:
    - add saj mqtt username/password to list of allowed logins in mosquitto config (this is required as the inverter is sending these credentials to the mqtt server and I don't think there is any way to change that):
      ```yaml
      logins:
        - username: "!secret mqtt_saj_user"
          password: "!secret mqtt_saj_password"
      ```
    - set `customize: true` and `folder: mosquitto` to mosquitto config
    - add `/share/mosquitto/saj_mqtt_bridge.conf` with following content to bridge saj app topic to still have access to the data from the saj app:
      ```
      connection saj-mqtt-bridge
      address 47.254.120.144
      remote_username <mqtt_saj_user>
      remote_password <mqtt_saj_password>
      topic saj/<serial>/realtime_data both
      topic saj/<serial>/realtime_dataPlus both
      ```
        - `address` contains the real ip of `mqtt.saj-solar.com`
        - `remote_username` contains the username used by saj to connect to `mqtt.saj-solar.com`
        - `remote_password` contains the password used by saj to connect to `mqtt.saj-solar.com`
        - topics are the list of topic we want to bridge (in my case it's `saj/<serial>/realtime_dataPlus`, but other firmwares might be using `saj/<serial>/realtime_data`)

With this setup, the app still gets the data, but I don't think remote control is still possible as I'm not bridging the data_transmission topic, so even if it's published on the remote broker, it won't be bridged as it's not part of the bridge topic config.

Only problem I can see now is that the ip address of `mqtt.saj-solar.com` can change and the bridge woudn't work anymore until the ip is updated in the bridge config.
If I would be able to change the mqtt server on the aio3 module directly (I know you can change it via the app, but I never tested it yet) and it actually also keeps that value after a reboot, then the adguard dns rewrite could be removed and in the bridge config, I could put `mqtt.saj-solar.com` as address instead. Then it should work also when the ip address is changing at saj side..
So I still need to check if the mqtt server can be changed and if it's kept after a restart of the aio3 module.

PS: I suppose you know the credentials for saj mqtt? If not, I can share in PM.

**Comment by h3llrais3r** [comment]
Let me know if the bridging works for you as well. 😉 

If you already poisened your dns locally, you can skip the adguard step for dns rewrite.

**Reply by walkjivefly** [reply]
Thanks for the detailed instructions. I will give them a go as soon as I can; hopefully this coming weekend.

I am aware that there is an infamously bad firmware version but not all the details. Have I perhaps misunderstood the info on the About screen and actually have v2.126 firmware and the _batteries_ have firmware version 1.361?
![img_20231107_140955](https://github.com/paolosabatino/saj-mqtt-ha/assets/17745270/3c3286b6-37ba-4347-989d-224d1aebc221)

Every single technical question I have asked of the installer and SAJ support has been ignored completely.

My inverter has (if the label is to believed) the eSolar WiFi2 dongle and although there is the option to reconfigure the destination address in the inverter; it is ignored and then reset to the default on reboot. So for my hardware/firmware it seems only DNS poisoning will work.

I do know the saj mqtt credentials; I did a bit of network sniffing soon after the inverter was installed and I was pissed off that SAJ or my installer reconfigured it remotely and wiped out the historical data in the inverter (which shouldn't have been there since it was supposed to be a brand new inverter! So arguably they haven't completely ignored all my questions after all).

I managed to get the inverter talking to mosquitto but couldn't get the integration to work. I found the first thing it does is subscribe to four topics:
    • saj/xxx/terminal_param_read
    • saj/xxx/terminal_param_write
    • saj/xxx/data_transmission
    • saj/xxx/update
and then starts publishing on
    • saj/xxx/update_rsp
    • saj/xxx/realtime_data
    • saj/xxx/information

But HomeAssistant wasn't happy. When it started up and tried to load the integration it complained about a (10 second) timeout expiry initialising the connection. Then after a short while it complained 3 times about an incorrect unit from a sensor and then complained every 70 seconds about the timeout initialising the connection.

It was simpler to abandon mqtt and revert to one of the web integrations instead. But I'd prefer to have realtime data, and no way for SAJ/installer to remotely mess with my configuration and data, if possible.

**Reply by h3llrais3r** [reply]
If you have wifi2 dongle, I suppose you still have the web integration where you could login remotely on the ip of the dongle?
In that case, I would probably use the web integration instead myself. 😉
With the AIO3 dongle however, this internal webserver is gone, and there is (according to me) no other way to integrate besides the modbus integration (with additional hardware, didn't try myself as I want to keep the dongle), the saj modbus over bluetooth app (https://github.com/sgsancho/saj_h1_s2_modbus_bluetooth, but didn't try myself yet) or mqtt (this integration).

**Reply by paolosabatino** [reply]
@walkjivefly if you have connection issues to MQTT by HomeAssistant and the integration, perhaps you should first configure HomeAssistant MQTT integration to talk with the broker.

In my setup I have mosquitto MQTT broker and HomeAssistant running on the same machine.
I also configured HomeAssistant to talk with MQTT to localhost (AFAIK HomeAssistant could also spawn its own MQTT broker).

The DNS is poisoned to induce the inverter to talk to my MQTT broker instead of the SAJ one.

The integration also only uses `data_transmission` and `data_transmission_rsp` messages: https://github.com/paolosabatino/saj-mqtt-ha/blob/20120d2cd51d29196b8a1925d3fcdcf872cdf179/custom_components/sajmqtt.py#L66

**Reply by h3llrais3r** [reply]
And be aware that the inverter is sending credentials to the mqtt broker. So you have to make sure that you allow the saj credentials on your local broker (without configuring those, I always got client not authorized error in the mosquitto logs)

**Reply by maegibbons** [reply]
Hi

Can I ask, how do you find the credentials that saj use for mqtt?

Krs

Mark

**Reply by h3llrais3r** [reply]
@maegibbons You can DM me on [h3llrais3r.github@gmail.com](mailto:h3llrais3r.github@gmail.com)

**Comment by h3llrais3r** [comment]
> My inverter has (if the label is to believed) the eSolar WiFi2 dongle and although there is the option to reconfigure the destination address in the inverter; it is ignored and then reset to the default on reboot.

That doesn't sound good, will have to try myself to see if it happens as well with the AIO3 module... 😢 

> I managed to get the inverter talking to mosquitto but couldn't get the integration to work. I found the first thing it does is subscribe to four topics:
> • saj/xxx/terminal_param_read
> • saj/xxx/terminal_param_write
> • saj/xxx/data_transmission
> • saj/xxx/update
> and then starts publishing on
> • saj/xxx/update_rsp
> • saj/xxx/realtime_data
> • saj/xxx/information

Hmm, I'm not seeing all these topics when I sniffed my network with tcpdump...
I only saw traffic on `saj/<serial>/realtime_dataPlus` topic...
Maybe the topic are different depending on the type of wifi module?
Maybe I didn't look good enough? How did you sniff the network for these topics?

**Reply by paolosabatino** [reply]
@walkjivefly Beware that you have H1-3.6K.S2 model; mine is H1.6K.S2 and @h3llrais3r is a H1.5K.S2; probably firmware version differs significantly!

**Comment by walkjivefly** [comment]
There is no webserver in my dongle. I'm using one of the integrations which talks to the SAJ server to get its data. Definitely not ideal and it needed tweaking to make it work. Messing around with a modbus hardware intervention was more effort than I wanted to make. mqtt was the middle ground and seemed like a good compromise which I'll try again with.

I used wireshark to capture what the dongle was up to - which also got me the credentials I needed to allow the dongle to connect to mosquitto after poisoning the DNS. I was surprised I had to sniff on the actual wifi rather than on the eth adapter where the mqtt server was, after all, the traffic ends up on the eth for the server to see it. 

I hadn't thought about the possibility of different capacity inverters using different firmwares. But that could explain why nothing I've tried works "out of the box". In which case it'll probably need more time spent on it than I'll have available this weekend. Ugh.

**Comment by Argon007** [comment]
interesting! I like the MQTT approach (if we get it working). I still want to use the SAJ mobile app but also get the real time info from the inverter in Home Assistant. So poisoning the DNS to a local MQTT broker in the LAN and forward the MQTT message to SAJ makes sense to me.

I have a H1-5K-S2 device with an AIO3 module
![Inverter](https://github.com/paolosabatino/saj-mqtt-ha/assets/40138759/e50acd0d-c7f5-4104-946e-ad5714706b48)

**Reply by h3llrais3r** [reply]
@Argon007 You have exactly the same setup as me, and I'm also from Belgium. 😉 
If you do as explained in the original post, it should work.
You can test it then by reading and parsing the data from the inverter by running the scripts from https://github.com/paolosabatino/saj-mqtt  
I'm really interested in your output to see if you get similar results as I. 😉

**Reply by Argon007** [reply]
Didn't had the time yet to try it out. I have Scarlet as ISP with a BBOX3, so first need to setup a DNS/DHCP server so that I can spoof the SAJ MQTT server. What did you use?

So to be clear, it works for you and you can capture the MQTT messages and read the current output in Watt for example? It would also mean that we can connect those MQTT messages and use it in Home Assistant in the Energy Dashboard. How fast are updates coming in via MQTT?

Thanks!
(ps: is there a way to send DM's directly to you in Github? I can't seem to find a way to do this)

**Reply by h3llrais3r** [reply]
I'm already running HA with adguard home addon. So with adguard Home you can rewrite the `mqtt.saj-solar.com` dns to your local broker instance (in my case it's the ip of HA as I'm running the mosquitto broker addon, not standalone).
Once you have that configured, the mqtt messages are posted to your local broker, but with the bridge config, also the messages for the app are still forwarded to the `mqtt.saj-solar.com` ip address (`47.254.120.144`)

This means that you can now use the scripts from https://github.com/paolosabatino/saj-mqtt to try to pull the data. These scripts have also been created by Paolo (owner of this repo) to test is standalone. 
If you test and you get meaningful results (would like to see if you get similar values as me) then you can add this custom integration to get the data in home assistant.

**Reply by BertBJ40** [reply]
h3llrais3r,
i'm a hardware men, programming is a lot harder for me.
I'm also from Belgium.  I want to control my SAJ home batteries.  In winter months charging at night.
i have an existing home assistant installing with mqtt broker for my home automation system.  Mqtt broker with username and password.
Is this a problem?
the rewrite of the mqtt.saj-solar.com dns can be done in home assistant/adguard?
do i need to install the scipts from https://github.com/paolosabatino/saj-mqtt or can i skip this?

Last but not least : can i communicate with you in dutch?

**Reply by Argon007** [reply]
Hi all,

I installed Adguard and can spoof the SAJ MQTT address now.
I've created a Telegram Group so that we can easily communicate and maybe help eachother to get it up and running. I also have some questions now about the next steps to take... :-)

You can join the group via this link: t.me/+e1OFuJBu49k1NDNk (append https:// at the beginning of the link).

**Reply by h3llrais3r** [reply]
> Last but not least : can i communicate with you in dutch?

@BertBJ40 Join the telegram group. There whe can communicate in Dutch.😉

**Comment by h3llrais3r** [comment]
> h3llrais3r, i'm a hardware men, programming is a lot harder for me. I'm also from Belgium. I want to control my SAJ home batteries. In winter months charging at night. i have an existing home assistant installing with mqtt broker for my home automation system. Mqtt broker with username and password. Is this a problem? the rewrite of the mqtt.saj-solar.com dns can be done in home assistant/adguard? do i need to install the scipts from https://github.com/paolosabatino/saj-mqtt or can i skip this?
> 
> Last but not least : can i communicate with you in dutch?

Not a problem, I'm also running mqtt broker with username/password. You'll just need to add the saj username/password to the list of logins in the mqtt broker config.
![image](https://github.com/paolosabatino/saj-mqtt-ha/assets/1945295/b4eda2b4-6284-46e5-aca0-7bf41da32f69)


The dns rewrite can be done in adguard with a single line (replace with your HA ip if you are running the mqtt broker addon in HA, or the IP of your mqtt standalone instance):
![image](https://github.com/paolosabatino/saj-mqtt-ha/assets/1945295/3a0ea06f-d30f-48d7-bb4d-2df6778fdf5f)

Scripts don't need to be installed themselves, you can just run them with python.
But you need to install python 3 and also install the python dependencies with pip once python is installed. (`pip install pymodbus paho-mqtt`)
Only change you have to make in the script itself is setting the username/password of your mqtt broker in the `readregs.py` 
![image](https://github.com/paolosabatino/saj-mqtt-ha/assets/1945295/0aa04e4e-5dd8-4417-978e-dc04083ce8d2)
Then you can run it like this (if you are on Windows):
`python readregs.py <your mqtt broker ip> <your inverter serial> 0x4000 0x100 | python parsedata.py -p`

Dutch is fine for me in direct communication. On github I rather keep it in English so everyone can follow. 😉

**Reply by stewepylon** [reply]
h3llrais3r, may you DM SAJ user and password please :)?

**Reply by h3llrais3r** [reply]
@stewepylon , you can DM me on h3llrais3r.github@gmail.com

**Reply by h3llrais3r** [reply]
Replied via DM. 😉

**Reply by miquelbotanch** [reply]
may you DM SAJ user and password please?

**Reply by h3llrais3r** [reply]
@miquelbotanch Please DM me on h3llrais3r.github@gmail.com so I can reply.

**Reply by erelke** [reply]
> @miquelbotanch Please DM me on [h3llrais3r.github@gmail.com](mailto:h3llrais3r.github@gmail.com) so I can reply.

I sent a DM with password req on 15. dec. Can you reply for me?

**Reply by h3llrais3r** [reply]
> > @miquelbotanch Please DM me on [h3llrais3r.github@gmail.com](mailto:h3llrais3r.github@gmail.com) so I can reply.
> 
> I sent a DM with password req on 15. dec. Can you reply for me?

Replied ;)

**Reply by h3llrais3r** [reply]
> @miquelbotanch Please DM me on [h3llrais3r.github@gmail.com](mailto:h3llrais3r.github@gmail.com) so I can reply.

Replied as well ;)

**Reply by Beufff** [reply]
Hi h3llrais3r,
I'm also looking for the SAJ user and password.
I already send you an email :)

Thank you very much !

**Reply by h3llrais3r** [reply]
> Hi h3llrais3r, I'm also looking for the SAJ user and password. I already send you an email :)
> 
> Thank you very much !

Replied ;)

**Comment by miquelbotanch** [comment]
Done!I’m looking forward to see how this works. I think is much cleaner than the elfin solution.Do you resend the information to the original broker ? Or do you use it isolated ?El 31 des. 2024, a les 20:22, h3llrais3r ***@***.***> va escriure:﻿

@miquelbotanch Please DM me on ***@***.*** so I can reply.

Replied as well ;)

**Reply by h3llrais3r** [reply]
Yes, data is still sent to the original broker. That's why we are bridging it, so you can still use the saj e-solar app. ;)

**Comment by miquelbotanch** [comment]
And what if they force an update using the broker, and the we loose the magic, because as far as I know, bringing is both ways so, exist the possibility that they force the upgrade and this propagate to our MQTT and then we loose all the magic.Enviat des del meu iPhoneEl 2 gen. 2025, a les 22:14, h3llrais3r ***@***.***> va escriure:﻿
Yes, data is still sent to the original broker. That's why we are bridging it, so you can still use the saj e-solar app. ;)

**Reply by h3llrais3r** [reply]
The bridge only bridges the 'realtime' data:
```
topic saj/<serial>/realtime_data both
topic saj/<serial>/realtime_dataPlus both
```
This is done to keep the saj app working.

Updates/changes are done via other topics.
Since they are not bridged, they cannot change anything anymore remotely, so they can't mess up your config. ;)
I know it's the case, because when I had some issues with my inverter, they contacted me that they couldn't do anything anymore remotely. So I had to temporarily remove the bridge to allow remote changes.

**Comment by walkjivefly** [comment]
Good afternoon, @h3llrais3r 

I'm finally revisiting this after being forced into action by SAJ desupporting the old website. I'm running a mosquitto MQTT broker in a container on the same Pi as HomeAssistant (also in a container). HA has the MQTT integration installed. I've poisoned the DNS, and MQTT requests from the eWifi2 dongle are being captured by mosquitto. I installed the saj_mqtt integration and have 77 sensors in HA from it. So far, so good!

The instructions above about how to set up the bridging were rather confusing. So I added the following to my mosquitto.conf
```
connection sajbridge
address 47.254.120.144
remote_username <username>
remote_password <password>
remote_clientid <inverter serial no>
topic saj/<inverter serial no>/realtime_data both
topic saj/<inverter serial no>/realtime_dataPlus both
try_private false
```
and the mosquitto log shows no errors connecting the bridge so I assume data is being propagated, but the SAJ portal at https://fop.saj-electric.com/saj and the new version at https://eop.saj-electric.com/index both show no data since the poisoning went into effect.

**Reply by h3llrais3r** [reply]
hi, I don't know about the portal... I'm only having the saj elekeeper app installed.
With my config (without `remote_clientid <serial>` and `try_private false`), I still see the data in the app... I never tried the website...
I'm not sure if those additional parameters have something to do with it...

EDIT: I've just tried to login on the portal and I'm seeing my data, so my config works.
Maybe 1 thing: what did you use for remote username/password?
Be aware that this is NOT your login, but the saj login used by the inverter (not your personal credentials).

**Reply by walkjivefly** [reply]
I was just being too impatient! The data is showing up in both the old and new portals now. 

The remote_clientid and try_private settings are to make the bridged connection to the cloud server look as much like an unbridged connection as possible. 

Now I'm "having fun" trying to work out why I have 2 or more copies of each entity in different statuses like "-", Disabled and Unmanageable, eg:
![image](https://github.com/user-attachments/assets/80ba56e1-0031-42a2-816a-d59aad43ce49)

Thank you, by the way, for the improvements you made to the integration.

**Reply by h3llrais3r** [reply]
Good to hear it's working. By the way, if you are just starting, i would suggest to try my newly rewritten integration which can be set up via the UI. 😉
EDIT: https://github.com/h3llrais3r/ha-saj-h1-mqtt

**Reply by maegibbons** [reply]
Hi.

I assume this is from your fork @

https://github.com/h3llrais3r/ha-saj-h1-mqtt

I will give it a go next week.

Krs

Mark

**Reply by walkjivefly** [reply]
Nice! Installed and running already. Now if only there was a safe, reliable way of transferring old sensor data!

**Reply by h3llrais3r** [reply]
Sorry, no way to migrate data... as sensors are renamed. Only way is renaming old sensors to new names manually before installing new integration.

**Reply by h3llrais3r** [reply]
> Hi.
> 
> I assume this is from your fork @
> 
> https://github.com/h3llrais3r/ha-saj-h1-mqtt
> 
> I will give it a go next week.
> 
> Krs
> 
> Mark

Yes

**Comment by stews87** [comment]
![IMG_8036](https://github.com/user-attachments/assets/e19b167b-71ad-4eaf-acd8-533620df68be)

**Comment by stews87** [comment]
configured everything but the bridge doesn't work

**Reply by walkjivefly** [reply]
How do you know the bridge doesn't work? Do you see messages in your broker that it connected to the SAJ server and the inverter connected to it? I thought mine wasn't working either, but it took about 2 hours before the numbers started showing up in the SAJ website.

Hopefully yours are there now as well.

**Comment by stews87** [comment]
File "/config/custom_components/saj_h1_mqtt/__init__.py", line 142, in async_setup_entry
    await async_first_refresh_on_mqtt_birth_message(hass, entry)
  File "/config/custom_components/saj_h1_mqtt/__init__.py", line 193, in async_first_refresh_on_mqtt_birth_message
    topic = hass.data[mqtt.DOMAIN].client.conf[mqtt.CONF_BIRTH_MESSAGE][mqtt.CONF_TOPIC]
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^

**Reply by h3llrais3r** [reply]
> File "/config/custom_components/saj_h1_mqtt/**init**.py", line 142, in async_setup_entry await async_first_refresh_on_mqtt_birth_message(hass, entry) File "/config/custom_components/saj_h1_mqtt/**init**.py", line 193, in async_first_refresh_on_mqtt_birth_message topic = hass.data[mqtt.DOMAIN].client.conf[mqtt.CONF_BIRTH_MESSAGE][mqtt.CONF_TOPIC] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^

@stews87 Can you please provide more details of the error? I think you missed the last part of it...
Also please provide the version of home assistant you are running.
Do you have the birth message configured in home assistant mqtt integration?
![image](https://github.com/user-attachments/assets/dacd6f45-e248-40b9-a643-01d3a6b29f5c)


To be sure it's caused by the last commit, you can always revert to this version (https://github.com/h3llrais3r/ha-saj-h1-mqtt/commit/787dd1356074a98d6c00fd574cdcbc018bf27cf9) and retry.

**Reply by h3llrais3r** [reply]
@stews87 I've just released a new version which should fix the problem.
Please update the custom repo via HACS and redownload the latest commit.
![image](https://github.com/user-attachments/assets/17dce3d2-67db-42f8-b144-7d1953fa3e9a)

Please let me know if it's fixed.

**Reply by stews87** [reply]
perfect!!! The integration now exposes 80 entities. The only problem is mosquitto keeps giving me error:2025-04-25 10:58:59: Connecting bridge saj-mqtt-bridge (47.254. 120.144:1883) 2025-04-25 10:58:59: Connection Refused: not authorised 2025-04-25 10:58:59: Client local.core-mosquitto.saj-mqtt-bridge closed its connection.
My country is on european node (italy) maybe change ip? Surely I do something wrong in the mosquitto configuration. please help me

**Reply by h3llrais3r** [reply]
It should be fine. I'm living in Belgium and using the ip I mentioned for 1.5 years now.
Are you using the right credentials in your broker config?
It's not your login, it's the SAJ mqtt login that you need.
If you don't have it, send me a mail on [h3llrais3r.github@gmail.com](mailto:h3llrais3r.github@gmail.com) and ping me here when it's sent.

**Reply by stews87** [reply]
ok, I had already sent him an email on the 24th you will find email sraf....@gmail.com. thanks

**Reply by h3llrais3r** [reply]
I've replied on your email with the saj credentials to be used in the broker config.

**Comment by stews87** [comment]
two days ago I could see the entities even if they are not available because it keeps giving me this error:
2025-04-25 10:58:59: Connecting bridge saj-mqtt-bridge (47.254.120.144:1883)
2025-04-25 10:58:59: Connection Refused: not authorised
2025-04-25 10:58:59: Client local.core-mosquitto.saj-mqtt-bridge closed its connection.
